File: tests/fixtures/java/basic/src/EdgeCases.java
package basic;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class EdgeCases {
    // Static initializer block (not a method)
    static {
        System.out.println("static init");
    }

    // Varargs parameter
    public String joinStrings(String separator, String... items) {
        return String.join(separator, items);
    }

    // Generic method
    public <T> List<T> wrapInList(T item) {
        return Arrays.asList(item);
    }

    // Generic return type
    public Map<String, Object> getMetadata() {
        return Map.of("key", "value");
    }

    // Annotation (custom)
    @Deprecated
    public void oldMethod() {}

    // Method with throws clause
    public void riskyMethod() throws Exception {
        throw new Exception("error");
    }

    // Inner class with methods
    public static class InnerHelper {
        public String help() {
            return "helping";
        }
    }

    // Anonymous class / lambda patterns
    public Runnable createTask() {
        return () -> System.out.println("task");
    }

    // Forward reference test: publicApi calls privateHelper defined LATER.
    // With two-pass extraction, this should resolve correctly (M1 fix).
    public String publicApi(String input) {
        return privateHelper(input);
    }

    private String privateHelper(String s) {
        return s.toUpperCase();
    }

    // this.method() call (M3 fix)
    public void callerUsingThis() {
        this.oldMethod();
    }
}

// Enum with methods (M4 fix)
enum Color {
    RED, GREEN, BLUE;

    public String display() {
        return this.name().toLowerCase();
    }

    public boolean isPrimary() {
        return this == RED || this == GREEN || this == BLUE;
    }
}

File: tests/fixtures/rust/basic/edge_cases.rs
use std::fmt;

// Enum with impl block
pub enum Color {
    Red,
    Green,
    Blue,
}

impl Color {
    pub fn display(&self) -> &str {
        match self {
            Color::Red => "red",
            Color::Green => "green",
            Color::Blue => "blue",
        }
    }
}

impl fmt::Display for Color {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.display())
    }
}

// pub(crate) visibility
pub(crate) fn crate_visible() -> bool {
    true
}

// Function with closure parameter
pub fn with_callback<F: Fn(i32) -> i32>(f: F) -> i32 {
    f(42)
}

// Lifetime parameters
pub fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() > y.len() { x } else { y }
}

// Result return type
pub fn parse_number(s: &str) -> Result<i32, String> {
    s.parse::<i32>().map_err(|e| e.to_string())
}

// n1 fix: renamed from UserService to EdgeService to avoid collision
struct EdgeService {
    name: String,
}

impl EdgeService {
    fn new(name: String) -> Self {
        Self { name }
    }

    // Self:: call test (M1 fix)
    pub fn create_default() -> Self {
        Self::new("default".to_string())
    }
}

// Static method call test — EdgeService::new defined above
pub fn static_call_test() {
    let _svc = EdgeService::new("test".to_string());
}

// Forward reference test (M2 fix): calls helper_b defined later
pub fn forward_ref_caller() {
    let _result = forward_ref_target();
}

fn forward_ref_target() -> i32 {
    42
}

// Closure call attribution test (n2)
pub fn closure_test() {
    let _result = with_callback(|x| x + 1);
}
